projects/packages/jetpack-mu-wpcom/src/features/launchpad/class-launchpad-task-lists.php

```diff
@@ -373,7 +373,30 @@ public function add_hooks_for_active_tasks( $task_list_id = null ) {
 			return;
 		}
 
-		$active_tasks = $this->get_active_tasks( $task_list_id );
+		$task_list_id = $task_list_id ? $task_list_id : get_option( 'site_intent' );
+		// Sites without a `site_intent` option will not have any tasks.
+		if ( ! $task_list_id ) {
+			return;
+		}
+
+		$task_list = $this->get_task_list( $task_list_id );
+		if ( empty( $task_list ) || ! isset( $task_list['task_ids'] ) ) {
+			return;
+		}
+
+		foreach ( $task_list['task_ids'] as $task_id ) {
+			$task_definition = $this->get_task( $task_id );
+			if ( isset( $task_definition['add_listener_callback'] ) && is_callable( $task_definition['add_listener_callback'] ) ) {
+				// We only need to know the built completion status if the task has an `add_listener_callback` property.
+				// Small optimization to not run `is_complete_callback` as often.
+				$task = $this->build_task( $task_definition );
+				if ( ! $task['completed'] ) {
+					call_user_func_array( $task_definition['add_listener_callback'], array( $task, $task_definition ) );
+				}
+			}
+		}
+
+		$active_tasks = wp_list_filter( $this->get_active_tasks( $task_list_id ), array( 'add_listener_callback' => true ) );
```

Comment:
@mattwiebe - Can you clarify the purpose of the changes above?

Reply:
Yes, it's an optimization over the previous implementation. This will reduce the number of times a `is_complete_callback` will ultimately be called, if it exists, when `$this->build_task()` gets called. We're getting to the same place, but a bit more efficiently.

Before:
1. Call `$this->get_active_tasks()`, which:
2. Gets the active task lists' tasks
3. Builds each of the task lists' tasks to determine `completed` status, which means firing `is_complete_callback` for any task that has one. Since this might be a bit expensive, I wanted to reduce that.
4. For each incomplete/active task (`completed = false`), add a matching `add_listener_callback` if the task definition has one.

After:
1. Get the active task list's task definitions
2. Filter by `add_listener_callback` - this will prevent ever having to build a task to determine `completed` status if it does not have the callback. We don't need to know that, so let's not calculate it.
3. For the list of task definitions with a listener callback, build each task to determine `completed`.
4. Each `completed = false` task in this list has its `add_listener_callback` called/added.

The last line is just a leftover line from my first implementation attempt, and totally wrecks all of the work I was trying to do here. Oops! Will remove.